### ISSUE TITLE
Theme support / dark theme

### ISSUE THREAD

**Issue by kjk** [sumatrapdfreader/sumatrapdf#3429]
This bug is to keep track of progress / bugs / ideas related to theming / dark theme.

Existing issues with bugs and ideas:
* https://github.com/sumatrapdfreader/sumatrapdf/issues/2224 Bookmark contents not readable with colors inverted
* https://github.com/sumatrapdfreader/sumatrapdf/issues/1586 "Frequently Viewed" Text does not change to reflect color settings
* https://github.com/sumatrapdfreader/sumatrapdf/issues/1275 Seperate theming from image inverting
* https://github.com/sumatrapdfreader/sumatrapdf/issues/898 Theme support plan
* https://github.com/sumatrapdfreader/sumatrapdf/issues/743 Theming improvements

Ideas:

1\. It would be good to re-use other people's theme e.g. Visual Studio's. like https://github.com/microsoft/vscode/blob/main/extensions/theme-quietlight/themes/quietlight-color-theme.json

That would give us access to tens of existing themes. It would require adopting (a subset of) their colors (just `colors` part, not the `tokenColors`).

We have JSON parser so could import the .json theme file directly.

Could easily ship the themes bundled in Sumatra, for easy access + allow 

If we do it, should happen before wide release to avoid changing things in the future.

**Comment by avorobey** [comment]
My guess is that >90% of what people want here is just the dark mode. Writing the code in terms of supporting >1 theme is nice; shipping many themes is nice; but having an easy way to toggle a reasonable dark mode is >90% use-cases in this area. So for example it would be counterproductive to ship a very nice set of 20 themes so that the user would need to hunt for "Dark" in the menu somewhere on the 17th place.

Right now we have two custom themes in the code ("Dark" and "darker") and I added a key combination to cycle through all three, but I think even cycling between 3 is slightly more confusing than just a way to toggle The Dark One. I've wavered between suggesting either 
- ship just two themes, light (default) and dark, and allow power-users to add however many customized themes via settings. 
- ship many themes, and allow choosing them via menu or cycling with a hotkey, but also have a distinguished Dark theme (maybe always the 2nd in the list after the default) and a separate hotkey to toggle Dark Mode on/off, which just moves between default and Dark themes.

**Comment by kjk** [comment]
Let me frame this as: define our own names for colors (what we do now) or re-use a subset of Visual Studio color names (a subset because it seems some colors do not apply).

If we adopt Visual Studio color names then it opens a path to future easy support / import of their themes. In the simplest case we could just say: copy & paste colors from the .json file to our settings format.

My thinking is: coming up with decent color scheme is hard and it would be good to be able to re-use the work other people  put in creating color schemes.

In any case: the color names we have currently are more of a haphazard evolution rather than a thoughtful design. The could use some scrubbing, especially some thought into minimizing the number of colors / settings, to simplify things.

Specifically, looking at Theme.h, maybe:
* TabCloseStyle could go
* TabTheme.closeCircleEnabled, closePenWidth could go
* MainWindowStyle.controlBackgroundColor could be same as MainWindowStyle.backgroundColor

**Comment by kjk** [comment]
Maybe we can simplify tab color customization.

In VS Code non-selected, non-higlighted tab doesn't draw the close button. On highlight (mouse hover) it shows the close button which is good enough to distinguish highlighted tab.

So minimalistic tab customization would be:
* text color, which probably could be the same as main text color and also used to draw the close X button
* background color
* highlighted background color

And one of the background colors could be the same as main background color which means we could reduce tab customization to one background color (not sure which should equal to main background color)

**Comment by kjk** [comment]
Another note: some UI, like About dialog, don't need to be themeable (because rarely used). VS Code doesn't bother theming about dialog.

**Comment by GitHubRulesOK** [comment]
Mainstream Windows Apps with tabs like Explorer 
![image](https://user-images.githubusercontent.com/12754913/230747161-f45f27ee-339b-4db6-8862-305870b73aa4.png)
![image](https://user-images.githubusercontent.com/12754913/230747287-d3770e8b-a004-4b0c-bdd6-58473501f5b5.png)

and Edge 
![image](https://user-images.githubusercontent.com/12754913/230747190-64be0b72-7a67-4881-84d0-27fb4dc172c0.png)
![image](https://user-images.githubusercontent.com/12754913/230747316-429cbc11-6475-4964-8713-b812e293f757.png)


have a fairly simple ethos. Explorer does have a stronger theme for one FILE tab but all the others merge into background with contrasted text which could on a simple basis be thresholded black or white or for those that gripe a custom value. thus tabs are transparent or light dark colour.

**Comment by GitHubRulesOK** [comment]
@kjk can I suggest that all the prior "Theme" issues like above and https://github.com/sumatrapdfreader/sumatrapdf/issues/2807 be closed if covered by this main issue so as to simply reference this one issue.

**Comment by gtumanyan** [comment]
Can we save the dark theme state so we can use it every time the program starts?

**Comment by GitHubRulesOK** [comment]
@gtumanyan  
as experimental there is not presently a constant to set, but a cyclic command
you can set it to a key like `i` is HOWEVER I suggest that say for Theme `t` is best not to mess with any invert abilities so

```
	[
		Cmd = CmdCycleTheme
		Key = t
	]

```

**Comment by george9816** [comment]
Please make the invert pdf option also work when the theme is changed. 

For me, the ideal setup is a black paper (inverted color) pdf with dark (gray color) theme for Sumatra.

**Comment by george9816** [comment]
Also, when switching theme change, scrollbars also should change color.

**Comment by GitHubRulesOK** [comment]
@george9816 you can see black scrollbars in my examples above but that is the Windows SYSTEM control of compiled Windows PAIN/E/T pick and replace with any color black word.

**Comment by paradoxicallist** [comment]
I don't know for sure but perhaps the following links might have valuable information re dark theme:
https://gist.github.com/rounk-ctrl/b04e5622e30e0d62956870d5c22b7017#dark-controls
https://github.com/derceg/explorerplusplus/issues/115#issuecomment-1598660527

**Comment by Hzbeta** [comment]
I propose the following three theme options, just like VS Code:

- **Light Theme**: This is a customizable, bright interface. Users can set it to a specific theme as per their preference.
- **Dark Theme**: This offers a customizable, darker interface, which can also be tailored to a specific theme according to user preference.
- **Follow System Theme**: This automatically aligns with the system's current theme to ensure consistency.

These options allow for quick toggling between two primary themes (light and dark), in addition to offering automatic syncing with the system theme.

Each theme comes with specific settings for all color aspects, encompassing both the PDF's BackgroundColor and TextColor.

**Comment by kjk** [comment]
I consider this done in latest pre-release (https://www.sumatrapdfreader.org/prerelease)

Use menu `Setting / Theme` or command palette (`Cmd + K`) and `Select next theme` command.

**Comment by fabiorzfreitas** [comment]
On general theme support, would it be possible to add a setting to apply the custom colors to "outside UI" (Title bar, File menu, Toolbar, etc.)?

If I understood it correctly, this is managed by the boolean `ThemeColorizeControls()` in `Theme.cpp`, right? I couldn't find a way to change this value on settings (aside from editing `Theme.cpp` and building from the source, as suggested).

I'd also like to draw your attention to the very popular [Dracula Theme](https://draculatheme.com/sumatra-pdf) (link to the SumatraPDF implementation). Alternatively, perhaps Dracula could be added to the themes list with Light, Dark and Darker!

**Comment by GitHubRulesOK** [comment]
@fabiorzfreitas 
that 3 year old  "Dracula theme" is just one of Gazillions ( 256x256x256x256x256x256x256x256x256 ) of personalisation's here try it yourself .
Replace SumatraPDF-settings.txt with the following and restart
```
FixedPageUI [
	TextColor = #f8f8f2
	BackgroundColor = #21222c
	SelectionColor = #44475a
	GradientColors = #0e0f12
```
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/c568850a-0a5b-4deb-9b16-b073d9ead72d)

**Comment by fabiorzfreitas** [comment]
> Replace SumatraPDF-settings.txt with the following and restart

This (and variations) is what I'd been trying for about one hour before asking here.

It worked neither on 3.5.2, nor 3.6.15958 with both portable and installer versions. I also thoroughly checked and removed every single file or registry entry before testing, there were no leftovers from previous installations.

The content section changes, but not the Title Bar, Toolbar and Tabs Bar.

---

BTW, I'm on Windows 11 23H2 Build 22631.3085, but [here are my full specs](https://web.archive.org/web/20240120192711/https://paste.rtech.support/selif/2s6d4hbe.html), gathered by /r/techsupport's tool.

---

> that 3 year old "Dracula theme"

On that, I can do the Math, thank you, but this age applies only to its SumatraPDF implementation among 365 other apps.

As of today, Dracula is the [4th most popular (full) theme extension](https://marketplace.visualstudio.com/search?target=VSCode&category=Themes&sortBy=Installs) for VSCode, and [106th over all categories](https://marketplace.visualstudio.com/search?target=VSCode&category=All%20categories&sortBy=Installs), with more than 6.5mi installs.

**Comment by GitHubRulesOK** [comment]
@fabiorzfreitas 
The dark title bar I show is Widows System Colour which worked on XP, Vista, Win7, Metro8, & 8.1, Windows 10, in all flavours and as shown the current pre-release build on current Windows 10.
 The difference you may be experiencing is that Windows 11 users say Windows theming seems to have been changed yet again (along with other negative system aspects such as printing).

Tabs and Toolbar are SumatraPDFs work area (controlled by optional "theme" but Title bar and Scrollbars are the Windows controlled frame, and can be Themed Globally by Windows Wide utilities.

**Comment by fabiorzfreitas** [comment]
> Tabs and Toolbar are SumatraPDFs work area (controlled by optional "theme") but Title bar and Scrollbars are the Windows controlled frame

As far as I can see, none of these are really controlled by Windows. Windows 11 allows choosing an "accent color", which will change the frame for the active window. SumatraPDF is not affected by this setting at all.

Also, I can't control Tabs and Toolbar as well.

While there is dedicated software for handling this system-wide, most of my apps control this behavior internally. MPC-BE, specially, even [displays it as a setting](https://i.imgur.com/UJSz3Kq.png).

For SumatraPDF, the built-in Dark and Darker themes are already able to change the whole window!

I suspect this is a result of the boolean I mentioned before, although I can't really find out because custom themes always are applied over the Light theme, the only one with `bool colorizeControls` set to false:

https://github.com/sumatrapdfreader/sumatrapdf/blob/7d2e69da4c1a661eef5126981c803072f7ac8f15/src/Theme.cpp#L59-L68

One could simply change line 110 to `true` and build from the source, but I think it would be much more feasible if this boolean could be assigned to something in the Custom Theme area.
